### PR TITLE
Improve reference & name resolution in ReplSeqMem

### PR DESCRIPTION
**Pull request by donggyukim**
This addresses #350, which is caused by bad reference resolutions for the same name memories. Do not merge until I add tests for this PR.

@zhemao This compiles your example, but not sure it's functionally working. Can you run the test with this branch?

### CODE REVIEW

File: src/main/scala/firrtl/passes/memlib/ReplaceMemMacros.scala [hunk 1]
```diff
@@ -148,7 +148,21 @@ class ReplaceMemMacros(writer: ConfWriter) extends Pass {
     }
   }
 
+  private type NameMap = collection.mutable.HashMap[(String, String), String]
```
Comment (azidar) [inline src/main/scala/firrtl/passes/memlib/ReplaceMemMacros.scala]: Comment - what are each of those strings?

File: src/main/scala/firrtl/passes/memlib/ReplaceMemMacros.scala [hunk 2]
```diff
@@ -148,7 +148,21 @@ class ReplaceMemMacros(writer: ConfWriter) extends Pass {
...
+  def constructNameMap(namespace: Namespace, nameMap: NameMap, mname: String)(s: Statement): Statement = {
```
Comment (azidar) [inline src/main/scala/firrtl/passes/memlib/ReplaceMemMacros.scala]: Comment - what does this function do?

**Comment by zhemao** [comment]
I've confirmed that this fixes the issue with running RocketChip unit tests in vsim.

**Comment by aswaterman** [comment]
@donggyukim can you address @azidar's changes so we can pull this into rocket-chip?

**Comment by shunshou** [comment]
How about the tests?

**Comment by donggyukim** [comment]
@shunshou I'll write them soon :)

**Pull request merged**